vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -282,67 +299,104 @@ def test_action_callback_function_return_value_valid(
+    ):
         action = Action(function=custom_action_function_mock_return())
         with pytest.raises(
             ValueError,
-            match="Number of action's returned elements \\(.?\\)"
-            " does not match the number of action's defined outputs \\(.?\\).",
+            match="Action function has not returned a list or similar but the action's defined outputs are a list.",
         ):
             action._action_callback_function(inputs={}, outputs=callback_outputs)
 
     @pytest.mark.parametrize(
         "custom_action_function_mock_return, callback_outputs",
         [
-            (
-                (namedtuple("Outputs", ["component_1_property"])("new_value")),
-                [],
-            ),
-            (
-                (namedtuple("Outputs", ["component_1_property", "component_2_property"])("new_value", "new_value_2")),
-                ["component_1_property", "component_2_property", "component_3_property"],
-            ),
+            (None, {"component_1_property": "value"}),
+            (False, {"component_1_property": "value"}),
+            (0, {"component_1_property": "value"}),
+            (123, {"component_1_property": "value"}),
+            ("", {"component_1_property": "value"}),
+            ([], {"component_1_property": "value"}),
+            ((), {"component_1_property": "value"}),
+        ],
+        indirect=["custom_action_function_mock_return"],
+    )
+    def test_action_callback_function_outputs_mapping_return_value_not_mapping(
+        self, custom_action_function_mock_return, callback_outputs
+    ):
+        action = Action(function=custom_action_function_mock_return())
+        with pytest.raises(
+            ValueError,
+            match="Action function has not returned a dictionary or similar "
+            "but the action's defined outputs are a dictionary.",
+        ):
+            action._action_callback_function(inputs={}, outputs=callback_outputs)
+
+    @pytest.mark.parametrize(
+        "custom_action_function_mock_return, callback_outputs",
+        [
+            ("", ["component_1_property"]),
+            ([], ["component_1_property"]),
+            ((), ["component_1_property"]),
+            ({}, ["component_1_property"]),
+            ("12", ["component_1_property"]),
+            ([1, 2], ["component_1_property"]),
+            ((1, 2), ["component_1_property"]),
+            ({"a": 1, "b": 2}, ["component_1_property"]),
+        ],
+        indirect=["custom_action_function_mock_return"],
+    )
+    def test_action_callback_function_outputs_list_return_value_length_not_match(
+        self, custom_action_function_mock_return, callback_outputs
+    ):
+        action = Action(function=custom_action_function_mock_return())
+        with pytest.raises(
+            ValueError,
+            match="Number of action's returned elements .+ does not match the number of action's defined outputs .+",
```

Comment:
Just out of curiosity, whast does the `.+` do?

Reply:
It matches 1 or more characters and we put it here because the original exception message is dynamically created like this:
`f"Number of action's returned elements ({len(return_value)}) does not match the number of action's defined outputs ({len(outputs)})."`